### DISCUSSION TITLE
Add state "buffering" to media_player

### DISCUSSION THREAD

**Discussion by emontnemery** [home-assistant/architecture#763] [Entity Models]
## Proposal:
Add new state `STATE_BUFFERING` to media_player

## Rationale:
When streaming media, the player may buffer for an extended period of time which is confusing if we can't show that to the user.
As an example, chromecast devices buffer some web radio streams for 15-30s.

With a new state `STATE_BUFFERING`, we can show a spinner or similar in the frontend while the player is buffering the media.

## Review of existing code:
Media players from integrations `cast`, `dunehd`, `openhome`, `plex`, `soundtouch` are already translating state "buffering" from the underlying API to `STATE_PLAYING`.

The `sonos` integration is displaying connection status `ZPSTR_CONNECTING` in the `media_title` attribute while buffering.

**Comment by frenck** [comment]
Sounds correct and the natural thing to do; with some integrations being able to support it as well; I like the UI feedback aspect of it.

Sounds good to me 👍

**Reply by amelchio** [reply]
I see this is already being implemented but I think the points in the @maxi1134 thread deserve consideration.

Looks like it could break a lot of automations with no easy workaround.

**Reply by MartinHjelmare** [reply]
I don't think we have implemented anything yet. Let's keep the discussion going until it's more settled.

**Reply by thecode** [reply]
Actually I think it is already too late: https://github.com/home-assistant/core/pull/70802

**Reply by MartinHjelmare** [reply]
Right, forgot about that. It's still only one integration using it though.

**Comment by maxi1134** [comment]
If I may, this makes some automation really hard to achieve.

Things like the volume of speakers being set when they get to "playing" from "off" become impossible since there is pretty much always the intermediate "buffering" state in between each song.

I can't think of a workaround for this since I have no way of seeing the state before "buffering"

Any ideas are appreciated!

**Reply by parautenbach** [reply]
Good point (I think). Should buffering rather be considered a special characteristic of the playing state and hence rather be a binary attribute?

**Reply by maxi1134** [reply]
That would make more sense to me!

**Reply by derandiunddasbo** [reply]
Also the intermediate buffering state essentially prevents any automation with a "for:" condition/trigger from working properly. I think, most users expect a media player to have a "stable" playing state, when it's obviously playing. Seing the state toggling between two states every couple of seconds is rather unexpected and overcomplicates any automations that need to deal with durations in triggers.

And last but not least, these frequent state changes unnecessarily fill the recorder database.

While I understand the use case for making buffering of a media player visible to users, the current implementation as the state of an entity is breaking a couple of other use cases.

**Reply by derandiunddasbo** [reply]
I have to add that I begin to appreciate, that I'm now able to see, if one of my cast devices is buffering. Especially for internet radio stations, which buffer for about 20 seconds before actually starting to play anything. But while the cast devices seem to stay in a stable playing state after initial buffering when playing internet radio, this is unfortunately not the case when streaming Spotify content to cast speakers. In this case the media_player entities are constantly toggling between "playing" and "buffering" states. Some of them every copule of minutes, but some of them even every couple of seconds. I don't know why there's a difference, as all my cast devices are connected to the same wifi.

To be clear: All my cast devices are playing Spotify content without any issues or interruptions, regardless of their frequency toggling between "playing" and "buffering" state. If I'd really want to show a spinner for every buffering event the cast player reports to HA, it would be quite annoying to watch the spinner appear and disappear this often without experiencing any actual issues with the audio being played. :-)

**Reply by elupus** [reply]
That is why my though was a separate attribute saying transitioning. I.e the new state is PLAYING, but it has not reached it fully yet. This would not break automations, and would also work with a transition between OFF and ON which may take some time. 

While it would not break automations, it might be a bit hard to automate on. So that is maybe a point against. Would likely need specific support in the automation engine to be easy to use. Ie. If you want to trigger on the started transition or the finished.

I suppose the automation engine could be extended to support intermediary states too.

**Comment by elupus** [comment]
Have additional bool attribute called "transiitioning" been condidered? Would solve off to on delays too.
